File: DesignPatterns/01_GOF23/DP_examples/C/02_Structural/Bridge/implementor.c
#include "implementor.h"

static const char* color_to_string(Color color) {
    switch (color) {
        case RED:
            return "red";
        case GREEN:
            return "green";
        case BLUE:
            return "blue";
        default:
            return "unknown";
    }
}

// 填充方法
static void Shape_fill(DrawAPI* self, Color color) {
    printf("Fill with %s\n", color_to_string(color));
}

// Circle
static void Circle_draw(DrawAPI* self) {
    printf("Draw a Circle\n");
}
DrawAPI* Circle_init(Circle* circle) {
    circle->base.draw = Circle_draw;
    circle->base.fill = Shape_fill;
    return &circle->base;
}

// Square
static void Square_draw(DrawAPI* self) {
    printf("Draw a Square\n");
}
DrawAPI* Square_init(Square* square) {
    square->base.draw = Square_draw;
    square->base.fill = Shape_fill;
    return &square->base;
}

// Rectangle
static void Rectangle_draw(DrawAPI* self) {
    printf("Draw a Rectangle\n");
}
DrawAPI* Rectangle_init(Rectangle* rectangle) {
    rectangle->base.draw = Rectangle_draw;
    rectangle->base.fill = Shape_fill;
    return &rectangle->base;
}

// BigCircle
static void BigCircle_draw(DrawAPI* self) {
    printf("Draw a Big Circle\n");
}
DrawAPI* BigCircle_init(BigCircle* bigCircle) {
    Circle_init(&bigCircle->base);
    bigCircle->base.base.draw = BigCircle_draw;
    return &bigCircle->base.base;
}

File: DesignPatterns/01_GOF23/DP_examples/Python/02_Structural/Bridge/Implementor.py
from abc import ABC, abstractmethod

# 颜色枚举
class Color:
    UNKNOWN = 0
    RED = 1
    GREEN = 2
    BLUE = 3

    @classmethod
    def to_string(cls, color):
        if color == cls.RED:
            return "red"
        elif color == cls.GREEN:
            return "green"
        elif color == cls.BLUE:
            return "blue"
        else:
            return "unknown"

class DrawAPI(ABC):
    @abstractmethod
    def draw(self):
        pass
    
    def fill(self, color):
        print(f"Fill with {Color.to_string(color)}")

class Circle(DrawAPI):
    def draw(self):
        print("Draw a Circle")

class Square(DrawAPI):
    def draw(self):
        print("Draw a Square")

class Rectangle(DrawAPI):
    def draw(self):
        print("Draw a Rectangle")

class BigCircle(Circle):
    def draw(self):
        print("Draw a Big Circle")
